hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/KafkaOffsetGen.java

```diff
@@ -148,9 +166,58 @@ public static OffsetRange[] computeOffsetRanges(Map<TopicPartition, Long> fromOf
         }
```

Comment:
Can we have the allocation done without needing to resplit. Allocation algorithm can be 

1. First allocate spark partitions uniformly across all TP, such that each spark partition reads min(events_remaining_in_tp, MAX_EVENTS_PER_BATCH/#tp). If no remaining data in a TP, we can skip it. Lets say, there are k spark partitions allocated this way. Lets say if "n" events are allocated, remaining = MAX_EVENTS_PER_BATCH - n
2. If min_partitions > k, then if there "x" TPs where each tp has remaining data (after previous step allocation) , allocate 1 spark partition to each such topics with min(events_remaining_in_tp, remaining/x). Now remaining = remaining - prev_allocated.
3. Repeat  step 2 until remaining == 0 or all topics are exhausted 

Let me know if this makes sense ?

Reply:
@bvaradar  I think the algorithm would not work well in data skew case. 
In data skew case, it will not divvy partition evenly. For example, given topic partitions "0:0->100, 1:0->500" and minPartitions=3, the algorithm will generate 2 ranges: "0:0->100, 1:0->200, 1:200->300", for the 2 ranges of partition 1, they are not divvied evenly. Given more skew partitions, it will be worse.
In the data skew case, resplit will generate even ranges for one TopicPartition. Because it will allocate ranges for topic partitions first, then based on the allocated ranges resplit into roughly minPartitions ranges.
Based on this and the complex of the resplit should be very small, I think resplit should be better.
How do you think?

Reply:
@waitingF : Considering your example ( here it assumes MAX_EVENTS_PER_BATCH = 300), I am not following your reasoning why it is not evenly distributed. Isn't it in your case, Partition 0 has only 100 events (0 -> 100) and so it got exhausted. In that case, you would need to fill the 3rd spark partition with events from TP 1.

Reply:
@bvaradar  I recoded the method based the algorithm, could you help check again? thanks a lot